pkg/provider/loadbalancer/fnutil/slice.go

```diff
@@ -53,21 +53,55 @@ func IsAll[T any](f func(T) bool, xs []T) bool {
 	return true
 }
 
-func IndexSet[T comparable](xs []T) map[T]bool {
-	rv := make(map[T]bool, len(xs))
+func PlanHashCode[DATA comparable](data DATA) DATA { return data }
+
+func IndexSet[DATA comparable](xs []DATA) *IndexSetWithComparableIndex[DATA, DATA] {
+	return NewIndexSetWithComparableIndex(PlanHashCode, xs)
+}
+
+type IndexSetWithComparableIndex[INDEX comparable, DATA any] struct {
+	hashCode func(data DATA) INDEX
+	data     map[INDEX]DATA
+}
+
+func NewIndexSetWithComparableIndex[INDEX comparable, DATA any](hashCode func(data DATA) INDEX, xs []DATA) *IndexSetWithComparableIndex[INDEX, DATA] {
```

Comment:
nit. I believe a single letter such as `I, T, or E` is clear enough for generic type names. Or `PascalCase` maybe.